erlang: refactor: build packages per Erlang/OTP version.

gleber opened the pull request:
This change introduces a fixpoint, which allows to do deep override when
building packages defined in pkgs/development/beam-modules/default.hex.
This allows to provide beam.packages.erlang{,R16,R17,R18,R19} which
contains the same packages built with different Erlang/OTP versions.

Top-level attribute beamPackages points at beam.packages.erlangR18, the
same applies to other top-level Erlang packages.

###### Motivation for this change

This is a continuation of https://github.com/NixOS/nixpkgs/pull/26381 based on https://github.com/NixOS/nixpkgs/pull/17240/

This does not yet generalize Elixir nor LFE interpreters.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip --against 9fb87f9c9d0eaeb97773e53e3740231d52b1104f"` while being at  9f7ae27361b71fe94af88bd165da47df8fe1c1a4
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review pr 26668"`
- [x] Tested execution of *many* binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

/cc @LnL7 @ericbmerritt @yurrriq @ankhers

gleber committed:
erlang: refactor: build packages per Erlang/OTP.

This change introduces a fixpoint, which allows to do deep override when
building packages defined in pkgs/development/beam-modules/default.hex.
This allows to provide beam.packages.erlang{,R16,R17,R18,R19} which
contains the same packages built with different Erlang/OTP versions.

Top-level attribute beamPackages points at beam.packages.erlangR18, the
same applies to other top-level Erlang packages.

TODO:
- beam.packages.erlang{R16,R17} is almost useless, since rebar/rebar3
  does not build using these versions;
- all packages in beam.packages which use buildMix are actually built
  with erlangR18;
- update documentation.

gleber commented:
from IRC:

<Ankhers> gleber_: Does that actually change anything? nix-review claims nothing changed.
<gleber_>: Ankhers: it does not change anything among top-level packages, but it does introduce many packages, which are not "exported"
<gleber_>: Ankhers: that's also one thing that I'd like to get feedback on. Is this really useful to have such "non-exported" packages defined at all? Should they be "exported"? Does this apply to only binaries, or to libraries too?

yurrriq commented:
Just glancing at this ok my phone for now, but does it enable easy building of arbitrary Hex packages as dependencies (and injecting them into `beam.pcakages` or whatever)? That's something I've wanted for a while.

gleber committed:
erlang: fix rebar3-nix-bootstrap in Erlang < R18.

gleber committed:
erlang: put Elixir/LFE into per-OTP packageset.

This makes beam.package.erlangR19.abnf to be actually built with R19,
instead of the default R18. It means that Elixir and LFE are provided in
two versions, one built with R18 and with R19.

Please note that Elixir does not build with R16 and R17 - trying to
access beam.packages.erlang{R16,R17}.elixir will throw an error.

gleber committed:
erlang: update documentation.

The documentation got a bit stale compared to actual contents of
nixpkgs. This commit focuses on updating existing docs, not on making
sure all details of beam packages are covered.

gleber commented:
@yurrriq I think it does provides some parts of what you want:
- `beam.packages.erlangRXX.buildHex` makes it easy to build a package from Hex;
- `beam.packages.erlangRXX` contains many packages from Hex (although this list is very outdated as you know).

There is no nice mechanism to do `beam.packages.erlang.override` to actually inject new packages into this package set, but it is probably not hard to add.

@LnL7 @ankhers @ericbmerritt This PR is ready to be reviewed.

ankhers commented on doc/languages-frameworks/beam.xml:
```diff
@@ -2,36 +2,85 @@
     <para>
       In this document and related Nix expressions we use the term
-      <emphasis>Beam</emphasis> to describe the environment. Beam is
+      <emphasis>Beam</emphasis> to describe the environment. BEAM is
```
We should probably change all `Beam` instances to `BEAM` if we are going to change a couple of them.

gleber replied:
I replace it only where it explicitly talks about BEAM the virtual machine, other than that Beam is used for stuff in Nixpkgs very consistently.

ankhers commented on doc/languages-frameworks/beam.xml:
```diff
@@ -2,36 +2,85 @@
+    </para>
+    <para>
+      We also provide <literal>beam.packages.erlang.callPackage</literal>, which
+      simplifies writng Beam package definitions, by injecting all packages from
```
Minor typo "writng".

gleber replied:
Done

ankhers commented on doc/languages-frameworks/beam.xml:
```diff
@@ -87,10 +137,11 @@ $ nix-env -f &quot;&lt;nixpkgs&gt;&quot; -iA beamPackages.ibrowse
-        <literal>buildRebar3</literal>. We use this function to make a
-        derivation that understands how to build the rebar3 project. For
-        example, the epression we use to build the <link
+        There is a Nix functional called <literal>buildRebar3</literal> (defined
```
I think this should be "There is a Nix function called..."

gleber replied:
Done

ankhers commented on doc/languages-frameworks/beam.xml:
```diff
@@ -181,6 +240,27 @@ $ nix-env -f &quot;&lt;nixpkgs&gt;&quot; -iA beamPackages.ibrowse
+          sha256 =
+              "08459823fe1fd4f0325a8bf0c937a4520583a5a26d73b193040ab30a1dfc0b33";
+          builder = buildMix;
+          beamDeps  = [ plug absinthe];
```
This is just stylistic, but maybe put a space between "absinthe" and the closing "]".

gleber replied:
Done

ankhers commented on doc/languages-frameworks/beam.xml:
```diff
@@ -264,18 +344,18 @@ let
-    <title>Building in a shell</title>
+    <title>Building in a shell (for mix projects)</title>
     <para>
       We can leveral the support of the Derivation, regardless of
```
I think this is meant to say "We can leverage the support..."?

gleber replied:
Done

LnL7 commented on pkgs/development/beam-modules/lib.nix:
```diff
@@ -12,6 +12,15 @@ rec {
+   let
+     parse = drv: (builtins.parseDrvName drv).version;
+   in builtins.replaceStrings ["B" "-"] ["." "."] (
+      if stdenv.lib.isString x
```
doesn't really matter, but `isString` is a builtin so you don't need stdenv.lib.

gleber replied:
I get this:

~/c/nixpkgs ❯❯❯ nix-build --show-trace -A beam.packages.erlangR18.elixir  /home/gleber/code/nixpkgs

error: while evaluating ‘versionAtLeast’ at /home/gleber/code/nixpkgs/lib/strings.nix:363:24, called from /home/gleber/code/nixpkgs/pkgs/development/beam-modules/default.nix:40:21:
while evaluating ‘versionOlder’ at /home/gleber/code/nixpkgs/lib/strings.nix:351:22, called from /home/gleber/code/nixpkgs/lib/strings.nix:363:29:
while evaluating ‘callPackageWith’ at /home/gleber/code/nixpkgs/lib/customisation.nix:95:35, called from /home/gleber/code/nixpkgs/pkgs/development/beam-modules/default.nix:6:9:
while evaluating ‘makeOverridable’ at /home/gleber/code/nixpkgs/lib/customisation.nix:54:24, called from /home/gleber/code/nixpkgs/lib/customisation.nix:99:8:
undefined variable ‘isString’ at /home/gleber/code/nixpkgs/pkgs/development/beam-modules/lib.nix:20:10

gleber replied:
Made it into builtins.isString instead

LnL7 commented on pkgs/top-level/beam-packages.nix:
```diff
@@ -44,22 +45,26 @@ rec {
-    lfe = callPackage ../development/interpreters/lfe { };
+    # Other Beam languages. These are built with `beam.interpreters.erlang`. To
+    # access for example elixir built with different version of Erlang, use
+    # `beam.packages.erlangR19.elixir`.
```
nice!

LnL7 reviewed (approved):
I have not played with it yet, but it looks good.

gleber commented:
Rebased, fixed all comments (by addressing them in appropriate commits, hence commit ids have changed).

ankhers commented:
`nox-review` passed on OSX.

gleber commented:
Current state of PR built successfully built on NixOS with sandboxes:

$ nix-shell -p nox --run "nox-review pr 26668"
...
Result in /run/user/1000/nox-review-982bk1l1
total 0
lrwxrwxrwx 1 gleber users 57 Jun 19 20:36 result -> /nix/store/lnxk39nmhg03pi5lniz3v8kih6vz0sxa-hex2nix-0.0.5
lrwxrwxrwx 1 gleber users 56 Jun 19 20:36 result-2 -> /nix/store/4npkcs2ds1dga0903mjbvj4ya3h3rq77-rebar3-3.3.2
lrwxrwxrwx 1 gleber users 59 Jun 19 20:36 result-3 -> /nix/store/mfdf6z2hjxpyk1rrnj0knpzxxdnrmlva-relx-exe-3.18.0
lrwxrwxrwx 1 gleber users 53 Jun 19 20:36 result-4 -> /nix/store/wx7bvpfimz0n68dkgjkai0kajxsx1h0a-lfe-1.2.1
lrwxrwxrwx 1 gleber users 56 Jun 19 20:36 result-5 -> /nix/store/n3j5kw7glc12g9gdvj83h39airmfinp1-rebar3-3.3.2
￼￼ 1￼

yurrriq commented:
This is working great for me, but I want to use `erlangR19` globally and I feel like this override is wrong or a bit clunky:

```nix
{
  nixpkgs.config.packageOverrides = pkgs: rec {
    beam = pkgs.beam // rec {
      interpreters = pkgs.beam.interpreters // {
        erlang = pkgs.beam.interpreters.erlangR19;
      };
      packages = pkgs.beam.packages // {
        erlang = pkgs.beam.packagesWith interpreters.erlang;
      };
    };
  };
}
```

Prior to the generalization, I just did this:

```nix
{
  nixpkgs.config.packageOverrides = pkgs: rec {
    erlang = pkgs.erlangR19;
  };
}
```

Maybe you can show me a better a way and maybe we can add that to the docs.

gleber commented:
@yurrriq That's a good piece of feedback! I am not sure how this can be done better.

I see two problems:
- if you override `beam.interpreters.erlang` it should switch `beam.packages.erlang` to the same version automagically (this is probably fixable by adding a fixpoint inside `beam` itself);
- the fact that you need to manually go first into `beam`, then into `interpreters` to do the override of `beam.interpreters.erlang` is a bit ugly - I suspect some sort of override function can be added inside `beam` to make this easier.

I will take a look sometime this week.

LnL7 commented:
I think would be useful to include erlang in the beamPackages attrset, then people can refer to that to get the correct version.

@yurrriq Note that overriding the interpreter like that has undesirable side effects, it causes top-level packages like couchdb to potentially break.

gleber commented:
@LnL7 I would be thankful if you provide guidance how to best structure ability to do the overrides on the levels where it is missing, as @yurrriq pointed out.

ankhers commented:
Is there anything else that needs to be done before this can be merged?

yurrriq commented:
@ankhers, I think we should at least add some documentation re: clean(er) overrides, including the caveat @LnL7 pointed out, and maybe clean that up in general, perhaps as @gleber mentioned.

yurrriq commented:
Re @LnL7's point: I agree. Really what I'm after is for my system-wide `erl` to be R19 and for LFE to be built with R19 too. I think I could instead include `beam.interpreters.erlangR19` in my `systemPackages` and override (or build a separate) LFE to use R19. If nothing else, I think some related examples in the docs would be highly beneficial.

ankhers commented:
@yurrriq I thought this would prevent the need for override? I thought that this change allowed you to install `beam.packages.erlangR19.lfe` to have it build on R19.

yurrriq commented:
Right, I think so. It'd be good to add a couple similar examples to the docs. /me checks the existing docs

yurrriq commented:
I've got some pedantic edits (spelling, grammar, continuity, etc) for the docs and can add some examples like I've mentioned. Otherwise I think this is good to go.

Thinking on it more, I don't think we need to support such dramatic global overrides as I've got on my system now (see above).

gleber commented:
Feel free to merge the PR as is - I think it already provides good value. I will continue refining it in various ways.

@yurrriq, sounds good. I will incorporate it into my work (regardless of the timing which PR gets merges/proposed first).

yurrriq commented:
@gleber: https://github.com/gleber/nixpkgs/pull/1

yurrriq committed:
Update BEAM docs

Improve beam docs:
 * correct spelling
 * update per pandoc changes
 * capitalize titles
 * capitalize BEAM throughout and use "the BEAM" when referring to the virtual machine.
 * tweak grammar and phrasing
 * reformat build-tools-rebar3 section
 * add more links
 * re-wrap <para>s

Also update <programlisting>s
* normalize whitespace
* don't double quote homepage
* use $ in all shell snippets

yurrriq commented:
We've merged in spelling/phrasing tweaks, including @ankhers's [suggestion](https://github.com/NixOS/nixpkgs/pull/26668#discussion_r122722745) of normalizing *BEAM* throughout, but I didn't add any examples as discussed. I should be able to get to that later today.

LnL7 commented on doc/default.nix:
```diff
@@ -26,7 +26,7 @@ pkgs.stdenv.mkDerivation {
     in ''
       {
-        pandoc '${inputFile}' -w docbook ${lib.optionalString useChapters "--chapters"} \
+        pandoc '${inputFile}' -w docbook ${lib.optionalString useChapters "--top-level-division=chapter"} \
```
What's this for?

yurrriq replied:
I got a warning that `--chapters` was deprecated in favor of `--top-level-division=chapter`.

yurrriq commented:
I got a warning that `--chapters` was deprecated in favor of `--top-level-division=chapter`.

LnL7 reviewed (approved):
I don't think improving the override is relevant enough to fix it now.

cstrahan commented on pkgs/development/beam-modules/default.nix:
```diff
@@ -1,16 +1,53 @@
+        cuter = callPackage ../tools/erlang/cuter {};
+        relxExe = callPackage ../tools/erlang/relx-exe {};
+      };
+in fix' (extends overrides packages)
```
@gleber For the sake of consistency, you might want to leverage [`makeExtensible`](https://github.com/NixOS/nixpkgs/blob/87b4a91fc4145e04ca028c08754bd144b8d1d02d/lib/fixed-points.nix#L70) here. It's presently used by both the Haskell package sets and `linuxPackages` sets.

The only change you'd need to make is replace this line with:

```
stdenv.lib.makeExtensible (extends overrides packages)
```

You could then remove the `overrides` argument, as you'll then be able to do `beam.packages.erlang.extend (self: super: { /* ... */ })`

Aside from aesthetics, I think there's a growing sentiment that "override" is a little bit overloaded, as it has historically been used to mean two different things:

* overriding the inputs into an expression, and
* overriding the packages within a package set

The use of "overriding" in the latter isn't exactly accurate, anyway -- you often want to "extend" the package set (and overriding can be seen as a subset of that concern). So the hope is that consistently using `.override` for modifying inputs and `.extend` for inputs will make things more intuitive, and having one way of going about it (`lib.makeExtensible`) will ensure that each package set gets it right, for free.

But, I'm not much involved in BEAM in Nixpkgs, so I figured I'd make the suggestion and leave the choice at your discretion :).

gleber replied:
Thank you for deep explanation, I am new to Nixpkgs so this kind of tribal knowledge is really hard for me to come by. I will include the suggested change into next PR.

A quick question: do you have any idea why `makeOverridable` does not have documentation? I could not find it neither in the code neither in the docs. Now, give that there are two distinct mechanisms which should be used for different situations, a documentation explaining both would be very useful.

cstrahan replied:
> Thank you for deep explanation

Sure thing -- thanks for the work on BEAM support :smile: 

> do you have any idea why makeOverridable does not have documentation?

Probably just oversight. I've opened a #26869 to (hopefully soon) resolve this.

gleber commented:
FYI https://github.com/NixOS/nixpkgs/pull/26813
